48 add basic logging

pdiercks opened the pull request:
see #48 for more info and future discussions how basic logging should be designed.

pdiercks commented on src/fenicsxconcrete/finite_element_problem/base_material.py:
```diff
@@ -49,34 +51,35 @@ def __init__(self,
-            level = self.p['log_level']
-            raise ValueError(f'unknown log level {level}')
+        self.logger.setLevel(self.p['log_level'])
+        # if self.p['log_level'] == 'DEBUG':
+        #     df.log.LogLevel(10)
+        #     logging.getLogger("FFC").setLevel(logging.DEBUG)
+        #     logging.getLogger("UFL").setLevel(logging.DEBUG)
+        #     logger.add(sys.stderr, level="DEBUG")
+        # elif self.p['log_level'] == 'INFO':
+        #     df.log.LogLevel(20)
+        #     logging.getLogger("FFC").setLevel(logging.INFO)
+        #     logging.getLogger("UFL").setLevel(logging.INFO)
+        #     logger.add(sys.stderr, level="INFO")
+        # elif self.p['log_level'] == 'WARNING':
+        #     df.log.LogLevel(30)
+        #     logging.getLogger("FFC").setLevel(logging.WARNING)
+        #     logging.getLogger("UFL").setLevel(logging.WARNING)
+        #     logger.add(sys.stderr, level="WARNING")
+        # elif self.p['log_level'] == 'ERROR':
+        #     df.log.LogLevel(40)
+        #     logging.getLogger("FFC").setLevel(logging.ERROR)
+        #     logging.getLogger("UFL").setLevel(logging.ERROR)
+        #     logger.add(sys.stderr, level="ERROR")
+        # elif self.p['log_level'] == 'CRITICAL':
+        #     df.log.LogLevel(50)
+        #     logging.getLogger("FFC").setLevel(logging.CRITICAL)
+        #     logging.getLogger("UFL").setLevel(logging.CRITICAL)
+        #     logger.add(sys.stderr, level="CRITICAL")
+        # else:
+        #     level = self.p['log_level']
+        #     raise ValueError(f'unknown log level {level}')
```
```suggestion
```
I would suggest to remove this. 
Not sure was `df.log.LogLevel(50)` does exactly, but I don't think it has the desired effect.
```
 ❯ python
Python 3.10.9 | packaged by conda-forge | (main, Feb  2 2023, 20:20:04) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import dolfinx.log as log
>>> log.LogLevel(50)
<LogLevel.???: 50>
>>> log.get_log_level()
<LogLevel.WARNING: -1>
```

I would suggest to not set anything here that is related to the components of fenicsx. IMO the user should have full control over each logger and what level is needed for the application.
See `tests/test_basic_logging.py` for how this could be done.

eriktamsen commented on tests/test_basic_logging.py:
```diff
@@ -0,0 +1,80 @@
+        dolfinx.log.LogLevel.ERROR,
+    ]
+    for lvl in dfx_levels:
+        print(f"Log level {lvl} has value: {lvl.value}.")
```
print statements in a test are not very useful in general, as the test are automated

pdiercks replied:
yes you are right. I included this for purpose of demonstration/documentation. the tests also show how one could set the level of different loggers. you could then run this test to get a list of supported values of `dolfinx.log.LogLevel`. 
But I can maybe better turn it into an assertion.

eriktamsen commented on tests/test_basic_logging.py:
Could you show an example, not only of how this works for dolfinx but how users would set different levels in an fenicsxconcrete example?
maybe testing a sample setup for all possible log levels using a pytest decorator for the different options as in [@pytest.mark.parametrize("setup", [CantileverBeam,](https://github.com/BAMresearch/FenicsXConcrete/blob/4a080611f039c99ce13d2be51b9cfae340d0bb3c/tests/experimental_setup/test_experimental_setups.py#L10)

pdiercks commented:
hi @eriktamsen ,
I added another test that shows how to set log levels for a fenicsxconcrete example. 
I added a logger to derived classes and the next step would be to define what should be logged.

eriktamsen commented on tests/test_basic_logging.py:
```diff
@@ -0,0 +1,112 @@
+        assert dolfinx.log.get_log_level().value == value
+
+
+@pytest.mark.parametrize("setup", [CantileverBeam,
```
I am not sure of testing this for all setups is necessary (was not what I meant in my other comment).
I was thinking more of a loop over all log levels, to show the different options.

pdiercks replied:
okay. I can add a loop over different levels if you like, but I am not sure if this would add any new information.

eriktamsen replied:
I would appreciate this. 
Again, this might be due to the fact, that I have not really looked into logging, but it would give me a nice overview of what levels exist and that they are working.

eriktamsen commented on tests/test_basic_logging.py:
```diff
@@ -0,0 +1,112 @@
+    problem = LinearElasticity(experiment, default_param)
+
+    for thing in [problem, experiment]:
+        assert thing.logger.getEffectiveLevel() == logging.INFO
```
is this checking the default?

pdiercks replied:
yes

eriktamsen commented on tests/test_basic_logging.py:
```diff
@@ -0,0 +1,112 @@
+
+    # set log level of each class instance
+    # using `set_log_levels`
+    my_levels = {problem.logger.name: logging.ERROR,
+                 experiment.logger.name: logging.DEBUG}
+    set_log_levels(my_levels)
+
+    assert problem.logger.getEffectiveLevel() == logging.ERROR
+    assert experiment.logger.getEffectiveLevel() == logging.DEBUG
```
maybe this is my lack of knowledge or me being unfamiliar with how logging works, but why do we now set different things for the problem and the experiment?

eriktamsen replied:
I guess we are talking about different use cases for this feature and maybe what I have in mind is something different.
I had the person in mind who might use one of our simulations without really knowing what is going on, not some involved developer who want to do detailed debugging.
The log_level would be a possibility to control the output of our package.

aradermacher replied:
Maybe it would be possible to add a possible (standard) logger set-up in the example, which is planned for the documentation.

pdiercks replied:
> I had the person in mind who might use one of our simulations without really knowing what is going on

in this case the default should be good?
But, I agree that it should be possible to set the level once for the entire package (all modules).
The use case I pointed out should also be possible and would make our life easier I guess.

pdiercks replied:
> Maybe it would be possible to add a possible (standard) logger set-up in the example, which is planned for the documentation.

yes, this is a better form to document the usage than this test.
maybe, we can do this once we have agreed on the basic structure?

eriktamsen reviewed (approved):
Thanks for adding this test. 
If this works I guess it is fine, so I approved it, however at least to me it does not feel very user friendly, or intuitive. @srosenbu, @aradermacher, @div-tyg, @danielandresarcones: what are your thoughts?
I don't think directly using the dolfin function set_log_level is optimal, as we are trying to get people to use our models without needing to think about dolfin.
Should we maybe implement our own function which wraps this? We can have a set_log_level both in the setup and the material to set the respective things manually.
In addition I think it could be useful to only set the log level once and not do it separately for setup and problem. I suggest to just pass another parameter for defining the overall log level. If it is set, the respective functions are called, if not it keeps the default.
I would be happy to implement this, if you agree.

aradermacher commented on src/fenicsxconcrete/experimental_setup/compression_cylinder.py:
```diff
@@ -99,6 +101,7 @@ def __init__(self, parameters: dict[str, pint.Quantity]=None):
         p.update(parameters)
 
         super().__init__(p)
+        self.logger = logging.getLogger("fenicsxconcrete.experimental_setup.compression_cylinder.CompressionCylinder")
```
I usually use "self.logger = logging.getLogger(__name__)" that should be the same here and is much shorter to write ;)

aradermacher replied:
I meant __name__ not bold

pdiercks replied:
you mean `__name__` ? yes that is a good idea. thanks :)

aradermacher replied:
_ _ name _ _

pdiercks replied:
it will be the name of the module and not the class though, but I guess this is fine.

aradermacher commented on src/fenicsxconcrete/experimental_setup/tensile_beam.py:
```diff
@@ -18,6 +23,7 @@ def __init__(self, parameters: dict[str, pint.Quantity]):
         default_p.update(parameters)
 
         super().__init__(default_p)
+        self.logger = logging.getLogger("fenicsxconcrete.experimental_setup.tensile_beam.TensileBeam")
```
see above

pdiercks commented:
hi @eriktamsen , @aradermacher 
what do you think about the latest changes? does this resolve the still open conversations?
I would add further examples in the documentation once we have fixed the basic structure.

eriktamsen commented on src/fenicsxconcrete/finite_element_problem/base_material.py:
```diff
@@ -55,34 +56,7 @@ def __init__(
-        else:
-            level = self.p["log_level"]
-            raise ValueError(f"unknown log level {level}")
+        self.logger.setLevel(self.p["log_level"])
```
Is there the same in the base experiment?
Then we would have a simple and in my opinion user friendly way of setting the log level for the whole problem.
Testing this would have been what I had in mind for the last test with the different levels. To give a simple example of how to set the different levels.

pdiercks replied:
> Is there the same in the base experiment?

I am not really satisfied with what I did here. I am thinking about a solution where the logging can be done on module-level, but also that each class can have a logger attribute.

>  Testing this would have been what I had in mind for the last test with the different levels. To give a simple example of how to set the different levels.

I think I had something like this in the first version .... :laughing:

eriktamsen replied:
I don't want to make this more complex and a bigger problem than it is. Feel free do decide that you know better and we will just see in the long run if this work for us. 
But while we are still discussing: What do you mean on a module level?
In practice, currently we have the setup and the material. although I don't see why we need to worry about setting different loggers for the different parts, but I am also not familiar with good practice in logging.
On that note, I have noticed, that the cylinder example uses some external module that has the default logging set to "info" which is a bit annoying. When running [this example](https://github.com/BAMresearch/FenicsXConcrete/blob/58-examples-in-documentation/docs/examples/cylinder.py) you will see what I mean. I would expect/suggest to make sure that external packages are set to the same logging level as module that calls them.

pdiercks replied:
> But while we are still discussing: What do you mean on a module level?

in the [tutorial](https://docs.python.org/3/howto/logging.html#advanced-logging-tutorial) they state module-level logging is a good convention. With `fenicsxconcrete` you might want to know where the log message came from.
If the loggers follow a naming hierarchy such as `name = self.__class__.__module__` and you do not explicitly set a level for this specific logger of a class, then the level by the `logger = logging.getLogger("fenicsxconcrete")` will be used.
In short, you can do
```python
from fenicsxconcrete import set_log_levels
set_log_levels({"fenicsxconcrete": logging.DEBUG})
```
to set the level for each logger within the package.

I think it would be nice if certain classes have a logger attribute. In the application you write you could use that to log important information.

> On that note, I have noticed, that the cylinder example uses some external module that has the default logging set to "info" which is a bit annoying. When running [this example](https://github.com/BAMresearch/FenicsXConcrete/blob/58-examples-in-documentation/docs/examples/cylinder.py) you will see what I mean. I would expect/suggest to make sure that external packages are set to the same logging level as module that calls them.

I guess this is the output from `dolfinx.io.gmshio.read_from_msh`. 
In general I would rather let the user set the log level for `dolfinx`, but the mesh info can be suppressed.

eriktamsen replied:
Maybe we need to move this discussion to the in person meeting. I still feel we are talking about different use cases.
And I think there is a way to allow for all the flexibility that you are trying to save while still making it easier for unexperienced people.
The way I want to design the interface for this package, is for people without having any idea what dolphinx is.
We are providing a package where we can develop features that are required within our group, but which need to be shared with other people who are not experts on FEM and are more or less able to use different python packages.
I agree that is is neccessary to have the possibility to define all the things you want but I dont want this to be the only and the default way.
Even for me, I would like to have preferably 3 settings: 1: no output unless something fails, 2: minimum output so that I know something is still going on internally, 3: maximum output when I need to understand whats happening
I dont want to then need to set different outputs or search how the definition of the logger is.

eriktamsen replied:
> So if I understand you correctly, the only thing is missing, is that the log level for fenicsxconcrete should also be set for dolfinx etc.?

Yes, that would be amazing, maybe likewise with the meshing package (or others that we use and that might have relevant output)

eriktamsen reviewed (approved):
If we assume that changing log level is only a development feature, then I am okay with the test example.
If we see a usecase where a user needs to/wants to change the log level, I would envision a more user friendly way of setting this. If I see correctly this is (partly) implemented.
I am also happy to merge this now and if we notice that we need a better way to use the log levels, make a new issue.

pdiercks commented:
hi @eriktamsen ,
I added `LogMixin` based on this [post](https://stackoverflow.com/questions/15780151/how-to-use-python-logging-in-multiple-modules). (see also reference therein on what a [mixin](https://stackoverflow.com/questions/533631/what-is-a-mixin-and-why-is-it-useful) is.)

by inheriting from `LogMixin` in the base classes, each child of those has a logger attribute. Due to the "namespace" (fenicsxconcrete) of the loggers we can easily set log levels for all loggers or individual ones (see tests/test_basic_logging.py).

IMO this would be sufficient for now. With that we have the basic structure to easily add new loggers and log messages to class methods.
I would leave the configuration of the logging (format, handlers) for later.

eriktamsen commented:
@pdiercks That sound quite good.
I still see the need for a simplified option to controll the output, see my comment above (maybe logging is the wrong term then), but I start to understand that this might be a seperate feature which we can deal with in a seperate issue afterwards.
I am happy for your suggestion to be merged, then we can all have the possbility to use it and based on that we can make better suggestions if needed.

pdiercks commented:
hi @eriktamsen ,
please have one last look on this. I extended `__init__.py` to enable the desired behaviour. 

Maybe, it makes more sense to put the `gmsh` stuff in another function and have the possibility to set more options? This might become relevant if the mesh creation process for the experiments gets more complicated in the future.
We could also directly initialize gmsh, but maybe this is better discussed in the future since there is currently no need.
